
services:
  app:
    image: ${NEXTCLOUD_IMAGE}:${NEXTCLOUD_TAG}
    container_name: nextcloud
    restart: unless-stopped
    depends_on:
      - db
      - redis
    env_file:
      - ../ops/.env.local
      - ./app.env
    environment:
      POSTGRES_HOST: ${NEXTCLOUD_DB_HOST}
      POSTGRES_DB: ${NEXTCLOUD_DB_NAME}
      POSTGRES_USER: ${NEXTCLOUD_DB_USER}
      POSTGRES_PASSWORD: ${NEXTCLOUD_DB_PASS}
      REDIS_HOST: ${NEXTCLOUD_REDIS_HOST}
      NEXTCLOUD_TRUSTED_PROXIES: traefik
      OVERWRITEPROTOCOL: https
    networks:
      - traefik
      - default
    volumes:
      - nextcloud_app:/var/www/html
      - ${HOMELAB_STORAGE_ROOT}/nextcloud/data:/var/www/html/data
    labels:
      - "traefik.enable=true"
      - "traefik.http.routers.nextcloud.rule=Host(`${NEXTCLOUD_HOST}`)"
      - "traefik.http.routers.nextcloud.entrypoints=websecure"
      - "traefik.http.routers.nextcloud.tls.certresolver=cloudflare"

      # This middleware is crucial for allowing large file uploads.
      # By default, Traefik might time out or reject them. This sets a 10GB limit.
      - "traefik.http.middlewares.nextcloud-uploads.buffering.maxRequestBodyBytes=10737418240"
      - "traefik.http.middlewares.nextcloud-uploads.buffering.memRequestBodyBytes=10737418240"

      - "traefik.http.routers.nextcloud.middlewares=secure-headers@docker,nextcloud-uploads"
      - "traefik.http.services.nextcloud.loadbalancer.server.port=80"
  db:
    image: ${NEXTCLOUD_DB_IMAGE}:${NEXTCLOUD_DB_TAG}
    container_name: nextcloud-db
    restart: unless-stopped
    env_file:
      - ../ops/.env.local
      - ./app.env
    environment:
      POSTGRES_DB: ${NEXTCLOUD_DB_NAME}
      POSTGRES_USER: ${NEXTCLOUD_DB_USER}
      POSTGRES_PASSWORD: ${NEXTCLOUD_DB_PASS}
    volumes:
      - nextcloud_db:/var/lib/postgresql/data
    networks:
      - default

  redis:
    image: ${NEXTCLOUD_REDIS_IMAGE}:${NEXTCLOUD_REDIS_TAG}
    container_name: nextcloud-redis
    restart: unless-stopped
    volumes:
      - nextcloud_redis:/data
    networks:
      - default

volumes:
  nextcloud_app:
  nextcloud_db:
  nextcloud_redis:

networks:
  traefik:
    external: true
  default:
    name: nextcloud_internal